### PR TITLE
Use webman/*.html and webman/api for OCaml.org manuals

### PR DESCRIPTION
**Pull request by shakthimaan**
Theis PR builds the HTML manual pages directly under webman/ and not webman/manual/ as preferred in https://github.com/ocaml/ocaml.org/issues/534. This will further allow the following URLs to be served directly under OCaml.org for easy navigation and reference:

```
  https://ocaml.org/manual/<VERSION>/*.html
  https://ocaml.org/manual/<VERSION/api/*.html
```

Reference:
* https://github.com/ocaml/ocaml.org/issues/534#issuecomment-1318570350

### CODE REVIEW

**Comment by Octachron** [comment]
I am not sure how this change would impact `ocaml.org` ? It is not like the files built in `webman/manual` or `webman` are published directly. 

Moreover, I have the impression that this PR would result in a different directory structure than the one described in the ocaml.org issue by mixing the manual html files and the `api` directory.

File: manual/src/html_processing/src/common.ml.in [hunk 1]
```diff
@@ -27,21 +27,23 @@ let ( // ) = Filename.concat
 
 let process_dir = Filename.current_dir_name
 
+let ocaml_version = "5.3.0"
```
Comment (Octachron) [inline manual/src/html_processing/src/common.ml.in]: The manual is shared between patch version, so `5.3` is probably better. And this version number is available as a configure time macro as `"@OCAML_VERSION_SHORT@"`

**Comment by shakthimaan** [comment]
The HTML manual documentation generated for 5.3.0 has been deployed for evaluation at https://shakthimaan.github.io/5.3.0/. (Eventually, we will replace shakthimaan.github.io with ocaml.org.)

> I am not sure how this change would impact ocaml.org ? It is not like the files built in webman/manual or webman are published directly.

Yes, the changes help in shorter URLs that can be easily remembered and shared. For example:

Current:      https://v2.ocaml.org/releases/5.1/htmlman/index.html
Suggested: https://ocaml.org/5.3.0/index.html

Current:       https://v2.ocaml.org/releases/5.1/api/Atomic.html
Suggested:  https://ocaml.org/5.3.0/api/Atomic.html

> this PR would result in a different directory structure than the one described in the ocaml.org issue by mixing the manual html files and the api directory.

Yes. This is in order to simplify the URLs as mentioned. The corresponding HTML files will be pushed to https://github.com/ocaml-web/html-compiler-manuals/ and served directly by ocaml.org.

Thanks for your valuable comments and feedback.

(P.S.: I am now part of the OCaml.org team and helping them with the documentation and tooling).

File: manual/src/html_processing/src/common.ml.in (hunk 1 above)
Comment (shakthimaan) [reply manual/src/html_processing/src/common.ml.in]: We want to make the manuals available for all OCaml releases, which includes minor versions as well. Hence, the x.y.z naming convention has been used for consistency.

Comment (Octachron) [reply manual/src/html_processing/src/common.ml.in]: Except that bugfix versions share the same features, manual and API that their main version. I am not convinced that it is useful to differentiate between them for the manual.

Comment (shakthimaan) [reply manual/src/html_processing/src/common.ml.in]: Okay. I will build using `"@OCAML_VERSION_SHORT@"` and shall update the PR. I now see the version handling at ocaml.org for version numbering at https://github.com/ocaml/ocaml.org/blob/86acab21cac4f887d2ce2b0d963813a78e3fa81d/src/global/url.ml#L42-L51. Thank you!

File: manual/src/html_processing/Makefile [hunk 2]
```diff
@@ -8,8 +8,9 @@ else
     DBG=quiet
 endif
 
-WEBDIR = ../webman
-WEBDIRMAN = $(WEBDIR)/manual
+VERSION=$(shell grep 'OCAML_VERSION_SHORT=' ../../../configure | awk -F'=' '{print $$2}')
```
Comment (Octachron) [inline manual/src/html_processing/Makefile]: The version number is available without shell invocation by including `Makefile.common` or `Makefile.config`.

File: manual/src/html_processing/Makefile [hunk 3]
```diff
@@ -1,3 +1,5 @@
+include ../../../Makefile.config
```
Comment (Octachron) [inline manual/src/html_processing/Makefile]: The compiler Makefile style for those definition is to define first a `ROOTDIR` directory:
```Makefile
ROOTDIR = ../../..
include $(ROOTDIR)/Makefile.common
```

File: manual/src/html_processing/Makefile [hunk 4]
```diff
@@ -8,8 +10,9 @@ else
     DBG=quiet
 endif
 
-WEBDIR = ../webman
-WEBDIRMAN = $(WEBDIR)/manual
+VERSION := $(OCAML_VERSION_MAJOR).$(OCAML_VERSION_MINOR)
+
+WEBDIR = ../$(VERSION)
```
Comment (Octachron) [inline manual/src/html_processing/Makefile]: As discussed offline, I would propose to keep `WEBDIR= ../webman` and use `WEBDIRMAN =$(WEBDIR)/$(VERSION)` as the output directory. Like this we keep the structure one directory by backend while still having the manual lives under a  ($VERSION) directory.

File: manual/src/html_processing/src/common.ml.in [hunk 5]
```diff
@@ -27,21 +27,23 @@ let ( // ) = Filename.concat
...
-let docs_maindir = web_dir // "manual"
+let docs_maindir = web_dir
 let docs_file = ( // ) docs_maindir
 
 (* Output for API *)
 let api_dir = web_dir // "api"
 
 (* How to go from manual to api *)
-let api_page_url = "../api"
+let api_page_url = "../" ^ ocaml_version ^ "/api"
```
Comment (Octachron) [inline manual/src/html_processing/src/common.ml.in]: Is there a reason why this path is not `./api`?
If this is the case, a comment would be nice.

File: manual/src/html_processing/src/common.ml.in [hunk 6]
```diff
@@ -105,6 +107,8 @@ let update_head ?(search = false) soup =
   |> append_child head;
   create_element "script" ~attributes:["src","navigation.js"]
   |> append_child head;
+  create_element "script" ~attributes:["src", "https://plausible.ci.dev/js/script.js"; "defer data-domain", "ocaml.org"]
```
Comment (Octachron) [inline manual/src/html_processing/src/common.ml.in:111]: I wouldn't mind a comment documenting what is the plausible script that we are linking here.

File: manual/src/html_processing/src/common.ml.in (hunk 5 above)
Comment (shakthimaan) [reply manual/src/html_processing/src/common.ml.in]: The manual here refers to `htmlman`. Without the above defined path, I get the following error:
```
Fatal error: exception Sys_error("../htmlman/./5.3/api/compilerlibref/index.html: No such file or directory")
```

Comment (Octachron) [reply manual/src/html_processing/src/common.ml.in]: Let's update the comment to reflect that point: `manual` ⇒ `htmlman manual`.

File: manual/src/html_processing/src/common.ml.in (hunk 6 above)
Comment (hannesm) [reply manual/src/html_processing/src/common.ml.in:111]: FWIW, I have a bad feeling to embed any random JavaScript from the Internet into the output of the manual -- it is then a burden for those who use an offline version of it. Instead, any ocaml.org specific stuff that is needed should be done in a post-processing step (or via a configure option or environment variable).

Comment (Octachron) [reply manual/src/html_processing/src/common.ml.in:111]: This script is already added in the postprocessing step for `ocaml.org` (notice the repertory name). I imagine we could add an offline flag to the post-processing script to skip this part of the postprocessing.

Comment (hannesm) [reply manual/src/html_processing/src/common.ml.in:111]: Oh, ok. Sorry for my lack of understanding about that. If I understand you correctly, the property that http://caml.inria.fr/distrib/ocaml-4.14/ocaml-4.14-refman-html.tar.gz (and future versions) allow offline browsing of the manual is fulfilled?

Comment (Octachron) [reply manual/src/html_processing/src/common.ml.in:111]: Indeed, those are the archive version that are not post-processed for `ocaml.org` and directly generated from the manual source.

File: manual/src/html_processing/src/common.ml.in (hunk 5 above)
Comment (Octachron) [reply manual/src/html_processing/src/common.ml.in]: The explanation is erroneous in fact, which explains why the path `htmlman/../5.3/api` doesn't correspond to anything.

The root issue here is that the `Process_manual.get_xfile` function filters the api index using the `../*` pattern when looking for files to add to the table of contents.

Using `../ocaml_version/api` rather than just `api` allow to still match this pattern by accident while keeping the correct url.

In other words, the simpler fix is to keep `api_page_url="api"` and fix `get_xfile` with
```diff
-             not (starts_with ".." rf) &&
+             not (starts_with "api" rf) &&
```

File: manual/src/html_processing/src/common.ml.in (hunk 6 above)
Comment (Octachron) [reply manual/src/html_processing/src/common.ml.in:111]: Maybe in term of comment :

> script for ocaml.org's instance of plausible analytics.

**Review by Octachron — APPROVED** [review]

**Comment by Octachron** [comment]
The PR looks good to merge, thanks! I would propose to add a change entry in the manual section (something like "better integration with ocaml.org"?) to keep track of the manual improvements.

**Comment by shakthimaan** [comment]
Should I create a new PR with just the changes along with a change entry ? Thank you!

**Comment by Octachron** [comment]
You can add the change entry in the changelog in a new commit in this PR. When merging, I will squash the PR into one commit since this is mostly a one feature PR with some back-and-forth and keeping the full history doesn't feel useful in those cases.
